File: android/app/src/main/java/music/mu/android/ui/PinSummary.kt
package music.mu.android.ui

import music.mu.android.PinManager
import mu.core.Scanner

/** 一組音軌的釘選狀態彙總（專輯詳情頁的釘選按鈕用；文案與 iOS 的 MuKit.PinSummary 一致）。 */
data class PinSummary(
    val total: Int,
    val done: Int,
    val pending: Int,
    val failed: Int,
) {
    companion object {
        fun of(tracks: List<Scanner.Track>, states: Map<String, PinManager.PinState>): PinSummary {
            val done = tracks.count { states[it.id] == PinManager.PinState.DONE }
            val pending = tracks.count {
                val s = states[it.id]
                s == PinManager.PinState.WANTED || s == PinManager.PinState.DOWNLOADING
            }
            val failed = tracks.count { states[it.id] == PinManager.PinState.FAILED }
            return PinSummary(tracks.size, done, pending, failed)
        }
    }

    val allDone: Boolean get() = total > 0 && done == total

    val shortLabel: String get() = when {
        total == 0 -> "無軌"
        allDone -> "已釘選"
        pending > 0 -> "釘選中 $done/$total"
        failed > 0 -> "重試釘選"
        else -> "釘選離線"
    }

    val fullLabel: String get() = when {
        total == 0 -> "無軌"
        allDone -> "已釘選（$total 軌，點擊取消）"
        done + pending > 0 -> "釘選中 $done/$total" + if (failed > 0) " · $failed 失敗" else ""
        failed > 0 -> "釘選失敗 $failed/$total（點擊重試）"
        else -> "釘選離線（$total 軌）"
    }
}

File: apple/CrateCore/Sources/CrateKit/PinSummary.swift
import CrateCore

/// 一組音軌的釘選狀態彙總（專輯/清單頭部的釘選按鈕共用；iOS/macOS 文案必須一致，
/// 因為 UI 測試斷言 `fullLabel` 的確切文字）。
public struct PinSummary {
    public let total: Int
    public let done: Int
    public let pending: Int
    public let failed: Int

    public init(tracks: [CrateCore.Scanner.Track], states: [String: PinManager.PinState]) {
        total = tracks.count
        done = tracks.filter { states[$0.id] == .done }.count
        pending = tracks.filter { states[$0.id]?.isPending == true }.count
        failed = tracks.filter { states[$0.id] == .failed }.count
    }

    public var allDone: Bool { total > 0 && done == total }

    /// 按鈕上的圖示名稱（SF Symbol）。
    public var symbolName: String {
        if allDone { return "checkmark.circle.fill" }
        if failed > 0 { return "exclamationmark.arrow.circlepath" }
        return "arrow.down.circle"
    }

    /// 按鈕短標籤。
    public var shortLabel: String {
        if total == 0 { return "無軌" }
        if allDone { return "已釘選" }
        if pending > 0 { return "釘選中 \(done)/\(total)" }
        if failed > 0 { return "重試釘選" }
        return "釘選離線"
    }

    /// accessibility / help 用完整狀態句。
    public var fullLabel: String {
        switch true {
        case total == 0:
            return "無軌"
        case allDone:
            return "已釘選（\(total) 軌，點擊取消）"
        case done + pending > 0:
            return "釘選中 \(done)/\(total)" + (failed > 0 ? " · \(failed) 失敗" : "")
        case failed > 0:
            return "釘選失敗 \(failed)/\(total)（點擊重試）"
        default:
            return "釘選離線（\(total) 軌）"
        }
    }
}
